test/agents/test_agent.py

```diff
@@ -0,0 +1,249 @@
+    tool_name, tool_input = agent._extract_tool_name_and_tool_input(pred)
+    assert tool_name == "Search" and tool_input == "Where was Jeremy McKinnon born"
+
+
+def test_extract_final_answer():
+    agent = Agent(prompt_node=MockPromptNode())
+
+    pred = "have the final answer to the question.\nFinal Answer: Florida"
+    final_answer = agent._extract_final_answer(pred)
+    assert final_answer == "Florida"
+
+
+def test_format_answer():
+    agent = Agent(prompt_node=MockPromptNode())
+    formatted_answer = agent._format_answer(query="query", answer="answer", transcript="transcript")
+    assert formatted_answer["query"] == "query"
+    assert formatted_answer["answers"] == [Answer(answer="answer", type="generative")]
+    assert formatted_answer["transcript"] == "transcript"
+
+
+@pytest.mark.integration
+@pytest.mark.parametrize("reader", ["farm"], indirect=True)
+@pytest.mark.parametrize("retriever_with_docs, document_store_with_docs", [("bm25", "memory")], indirect=True)
+@pytest.mark.skipif(
+    not os.environ.get("OPENAI_API_KEY", None),
+    reason="Please export an env var called OPENAI_API_KEY containing the OpenAI API key to run this test.",
+)
+def test_agent_run(reader, retriever_with_docs, document_store_with_docs):
+    search = ExtractiveQAPipeline(reader, retriever_with_docs)
+    prompt_model = PromptModel(model_name_or_path="text-davinci-003", api_key=os.environ.get("OPENAI_API_KEY"))
+    prompt_node = PromptNode(model_name_or_path=prompt_model, stop_words=["Observation:"])
+    counter = PromptNode(
+        model_name_or_path=prompt_model,
+        default_prompt_template=PromptTemplate(
+            name="calculator_template",
+            prompt_text="When I give you a word, respond with the number of characters that this word contains.\n"
+            "Word: Rome\nLength: 4\n"
+            "Word: Arles\nLength: 5\n"
+            "Word: Berlin\nLength: 6\n"
+            "Word: $query?\nLength: ",
+            prompt_params=["query"],
+        ),
+    )
+
+    agent = Agent(prompt_node=prompt_node)
+    agent.add_tool(
+        Tool(
+            name="Search",
+            pipeline_or_node=search,
+            description="useful for when you need to answer "
+            "questions about where people live. You "
+            "should ask targeted questions",
+        )
+    )
+    agent.add_tool(
+        Tool(
+            name="Count",
+            pipeline_or_node=counter,
+            description="useful for when you need to count how many characters are in a word. Ask only with a single word.",
+        )
+    )
+
+    result = agent.run("How many characters are in the word Madrid?")
+    assert any(digit in result["answers"][0].answer for digit in ["5", "6", "five", "six"])
+
+    result = agent.run("How many letters does the name of the town where Christelle lives have?")
+    assert any(digit in result["answers"][0].answer for digit in ["5", "6", "five", "six"])
```

Comment:
Same for this one, not sure who Christelle is but I'd expect the name of a town to be consistent. 😅

Reply:
This `Counter` tool is probably not well suited for the tests. Counting characters doesn't work well with the LLMs it seems. Probably because they work on token level. I'll see whether I find a better tool for the tests. Once some tools are implemented in Haystack, we can use them here. Consider it a placeholder until we have proper tools.

Reply:
Good for me then. We can rework that in the future. 👍